### PR TITLE
Fix bit const

### PR DESCRIPTION
**Pull request by rsetaluri**
See https://github.com/cdonovick/peak/issues/147 for more discussion.

Also related is #651. Honestly I think we need to rework a few things in the type system; not w/ respect to type hierarchy or relations or anything. Just w.r.t. interfaces and construction. Right now it seems very ad-hoc and not super robust (e.g. see `Array::__init__`). Discussion re: https://github.com/cdonovick/peak/issues/146 also comes into play, regarding protocol types.

### CODE REVIEW

File: magma/bits.py [hunk 1]
```diff
@@ -94,7 +94,7 @@ def __init__(self, *args, **kwargs):
             self.ts = args[0].ts[:]
             # zext for promoting width
             for i in range(len(self) - len(args[0])):
-                self.ts.append(GND)
+                self.ts.append(Bit.GND)
```
Comment (leonardt) [inline magma/bits.py:97]: Should this be Bit(0) for consistency with elsewhere?

**Comment by rsetaluri** [comment]
See garnet flow passing: https://buildkite.com/stanford-aha/aha-flow/builds/1226

File: magma/bits.py (hunk 1 above)
Comment (rsetaluri) [reply magma/bits.py]: Fixed

**Comment by rsetaluri** [comment]
@leonardt Agree with everything. It looks like there are 3 distinct issues at play; here they are along with my suggested solution:

1. Type constructors vs. type conversions, e.g. `Bits()` vs `bits()`. I think type constructors should be very simple, and be able to build values only from python POD's. I can write up a set of rules we can implement. Additionally, I think type conversions should all be from one magma type to another. That way these 2 mechanisms are completely separate. Ideally, the lower case "conversion" functions would be called "casts" or "reinterprets" but NBD. The only issue here is backwards compatibility since these constructors/conversions are used all over the place. May require a big overhaul.
2. Whether or not `GND` or `VCD` should be special named bits. I don't think they should be. We should just keep a `value` member which is `None` in the non-const case, and `0` or `1` otherwise. (See https://github.com/phanrahan/magma/blob/ee8da8eb670e9d6bf5878999b752f48eb0b139b3/magma/bit.py#L122-L137; already kind of doing this but seems inconsistent.) This somewhat ties into the next point, when it comes to checking equality...
3. The `==` operator can not be used for checking in the meta-language for equality to 0/1 because `==` instances a circuit. Well, this is only partly true because it does so for `Bit`, `Bits`, `SInt`, and `UInt` *always*, and *sometimes* for `Array`, `Tuple`, `Clock`, etc. (see https://github.com/phanrahan/magma/blob/ee8da8eb670e9d6bf5878999b752f48eb0b139b3/magma/array.py#L175-L178). It is a bit weird for `eq` to return a python value T/F sometimes, and a `Bit` other times. It makes it hard to have generic code without checking for these 2 cases. Instead, I think `==` operator on all types should return a Bit; for nested/complex types, we just dispatch on the subtypes. If you want to check that a bit is 0/1, you should do `x.const()` then `x.const_value()` which returns T/F (or 0/1). If you really want to get a python value for an equality op,  we can create a `def is(self, other)` which recursively does like `self is other or all(x.is(y) for x, y in zip(self, other))`.

What do you think?  Maybe I can start an issue to document all this type stuff and nail it down.

**Comment by rsetaluri** [comment]
...but for now will merge to fix the present bug

**Pull request merged**